src/crush/CrushWrapper.cc

```diff
@@ -1817,6 +1821,144 @@ int CrushWrapper::update_device_class(int id,
   return 1;
 }
 
+// Recursively get all only-child shadow parents up to the root
+// based on the given item id.
+// Return -EBUSY if any parent has more than one children.
+// Return -ENOENT if a specific parent can not be found.
+// Return 0 on success.
+int CrushWrapper::get_immediate_only_child_shadow_parents(CephContext *cct,
+                                                          int id,
+                                                          list<int> *parents)
+{
+  assert(parents);
+  parents->clear();
+  while (_search_item_exists(id)) {
+    int parent;
+    int r = get_immediate_parent_id(id, &parent, PARENT_SHADOW);
+    if (r < 0) {
+      ldout(cct, 0) << __func__ << " unable to get immediate shadow parent of"
+                    << " item id = " << id << dendl;
+      return -ENOENT;
+    }
+    crush_bucket *b = get_bucket(parent);
+    if (IS_ERR(b)) {
+      ldout(cct, 0) << __func__ << " unable to locate parent id = "
+                    << parent << dendl;
+      return -ENOENT;
+    }
+    if (b->size != 1) {
+      assert(b->size > 0);
+      ldout(cct, 0) << __func__ << " parent id = " << parent
+                    << " have " << b->size << " children,"
+                    << " returning" << dendl;
+      return -EBUSY;
+    }
+    parents->push_back(parent);
+    id = parent;
+  }
+  return 0;
+}
+
+int CrushWrapper::remove_device_class(CephContext *cct, int id, ostream *ss)
+{
```

Comment:
Yeah sure.. but should we do --show-shadow on 'osd crush tree' instead of 'osd tree'?  It seems more appropriate there (and won't mix with the up/down/in/out/destroyed stuff).